kernel/frame_allocator/src/lib.rs

```diff
@@ -382,24 +572,61 @@ impl AllocatedFrames {
             return Err(other);
         }
 
-        // ensure the now-merged AllocatedFrames doesn't run its drop handler and free its frames.
+        // ensure the now-merged Frames doesn't run its drop handler
         core::mem::forget(other); 
         Ok(())
     }
 
-    /// Splits this `AllocatedFrames` into two separate `AllocatedFrames` objects:
+    /// Splits up the given `Frames` into multiple smaller `Frames`.
+    /// 
+    /// Returns a tuple of three `Frames`:
+    /// 1. The `Frames` containing the requested range of frames starting at `start_frame`.
+    /// 2. The range of frames in the `self` that came before the beginning of the requested frame range.
+    /// 3. The range of frames in the `self` that came after the end of the requested frame range.
+    /// 
+    /// If `start_frame` is not contained within `self` or `num_frames` results in an end frame greater than the end of `self`,
+    /// then `self` is not changed and we return (self, None, None).
+    pub fn split(
+        self,
+        start_frame: Frame,
+        num_frames: usize,
+    ) -> (Self, Option<Self>, Option<Self>) {
+        if (start_frame < *self.start()) || (start_frame + (num_frames - 1) > *self.end()) || (num_frames == 0) {
+            return (self, None, None);
+        }
```

Comment:
it's not really testing for overlap, more to see if the frame range parameter is contained within self.

Reply:
isn't that the same as what `FrameRange::overlap()` does? at least the logic looks quite similar to me

Reply:
It's one case of an overlap where a range is not the subset of another. The expectation of the split function is that the frames that are passed are a subset of self which is another case of an overlap.

e.g. take 2 ranges [0,3] and [1,4]. The overlap function will return Some() to indicate there's an overlap whereas our check will return false.

Reply:
where self = [0, 4] and the frame range parameter is [1,3]

Reply:
oh ok, so this is `contains()` then, right? sorry i misunderstood.

basically i'm trying to figure out how we can refactor and reuse this logic elsewhere, perhaps within `RangeInclusive` itself? e.g.:
```rust
impl RangeInclusive {
    pub fn contains_range(&self, other: &RangeInclusive) -> bool { ... }
```

Reply:
yeah we can do a contains function, but I guess do it at the `FrameRange` and `PageRange` level since that's where overlap is defined as well, and as you mentioned in another place, we might end up changing the `RangeInclusive` we use in the future.